sign/eddilithium3/eddilithium.go

```diff
@@ -0,0 +1,210 @@
+// eddilithium3 implements the hybrid signature scheme Ed25519-Dilithium3.
+package eddilithium3
+
+import (
+	"crypto"
+	cryptoRand "crypto/rand"
+	"errors"
+	"io"
+
+	"github.com/cloudflare/circl/internal/shake"
+	"github.com/cloudflare/circl/sign/dilithium/mode3"
+	"github.com/cloudflare/circl/sign/ed25519"
+)
+
+const (
+	// Size of seed for NewKeyFromSeed
+	SeedSize = mode3.SeedSize // = ed25519.SeedSize = 32
+
+	// Size of a packed PublicKey
+	PublicKeySize = mode3.PublicKeySize + ed25519.PublicKeySize
+
+	// Size of a packed PrivateKey
+	PrivateKeySize = mode3.PrivateKeySize + ed25519.PrivateKeySize
+
+	// Size of a signature
+	SignatureSize = mode3.SignatureSize + ed25519.SignatureSize
+)
+
+// PublicKey is the type of an EdDilithium3 public key.
+type PublicKey struct {
+	e ed25519.PublicKey
+	d mode3.PublicKey
+}
+
+// PrivateKey is the type of an EdDilithium3 private key.
+type PrivateKey struct {
+	e ed25519.KeyPair
+	d mode3.PrivateKey
+}
+
+// GenerateKey generates a public/private key pair using entropy from rand.
+// If rand is nil, crypto/rand.Reader will be used.
+func GenerateKey(rand io.Reader) (*PublicKey, *PrivateKey, error) {
+	var seed [SeedSize]byte
+	if rand == nil {
+		rand = cryptoRand.Reader
+	}
+	_, err := io.ReadFull(rand, seed[:])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pk, sk := NewKeyFromSeed(&seed)
+	return pk, sk, nil
+}
+
+// NewKeyFromSeed derives a public/private key pair using the given seed.
+func NewKeyFromSeed(seed *[SeedSize]byte) (*PublicKey, *PrivateKey) {
+	var seed1 [32]byte
+	var seed2 [32]byte
+
+	h := shake.NewShake256()
+	_, _ = h.Write(seed[:])
+	_, _ = h.Read(seed1[:])
+	_, _ = h.Read(seed2[:])
+	dpk, dsk := mode3.NewKeyFromSeed(&seed1)
+	epair := ed25519.NewKeyFromSeed(seed2[:])
+
+	return &PublicKey{epair.GetPublic(), *dpk}, &PrivateKey{*epair, *dsk}
+}
+
+// SignTo signs the given message and writes the signature into signature.
+// It will panic if signature is not of length at least SignatureSize.
+func SignTo(sk *PrivateKey, msg []byte, signature []byte) {
+	mode3.SignTo(
+		&sk.d,
+		msg,
+		signature[:mode3.SignatureSize],
+	)
+	esig, _ := sk.e.SignPure(
+		msg,
+	)
+	copy(signature[mode3.SignatureSize:], esig[:])
+}
+
+// Verify checks whether the given signature by pk on msg is valid.
+func Verify(pk *PublicKey, msg []byte, signature []byte) bool {
+	if !mode3.Verify(
+		&pk.d,
+		msg,
+		signature[:mode3.SignatureSize],
+	) {
+		return false
+	}
+	if !ed25519.Verify(
+		pk.e,
+		msg,
+		signature[mode3.SignatureSize:],
+	) {
+		return false
+	}
```

Comment:
Yes, using timing the attacker can see which of the signatures fails.  Dilithium and Ed25519 signature verification aren't constant time either themselves: they leak information about *which* signature is being verified, just like most signature schemes.  (But it doesn't leak information about the private key as it obviously doesn't come into play at all during verification.)

Reply:
Got it! But I was thinking that maybe here we can always check `!mode3.Verify && !ed25519.Verify` at the end of the function and then return false..

Reply:
The compiler will turn that into just checking mode3.Verify and then returning false immediately when that is false. ( https://godbolt.org/z/5HaWu7 )

To prevent this compiler optimisation we could add a variant of verify that returns uint64s instead of bools and use `mode3.VerifyI() | ed25519. VerifyI()`.  I'm not sure whether such a complication of the API is worth it.

Reply:
Oh, right. Thanks for letting me now @bwesterb . Let's keep it the way you did it ;)

Reply:
Or we could just use `ConstantTimeByteEq`, casting the inputs to `byte`s and output to a `bool`? I don't see how this complicates the API, given that this is internal to `Verify`.

Reply:
That might also work, but it doesn't change the fact that signature verification of Ed25519 and Dilithium aren't constant-time themselves.  Even if we make this check constant time, we'll still leak exactly the same fact because Dilithium and Ed25519 verification aren't constant time.

Reply:
Yep, I was going to suggest using `subtle`, as @chris-wood recommends.

Reply:
> Let's keep it the way you did it
 
 Agree

Reply:
Again, using constant time operations here does not do anything to the overall behaviour of the function as ed25519.Verify and dilithium3.Verify are very much not constant time.

Reply:
Is it possible for those functions to be done in constant time? If so, I see no reason to make this constant time on the assumption that these two functions might later change. The cost is basically nothing.

Reply:
It is possible to create constant-time versions of dilithium3.Verify and ed25519.Verify, but they would be significantly slower.  The reason I don't like making this function here constant-time, is that it is a very clear statement "here is something special: here we should be constant time", whereas in fact that is useless action.  If it would be commonplace for signature schemes to have constant-time verification, I would think differently to prevent copy-paste errors, but that is not the case: signature schemes generally don't have constant-time verification operation.

Reply:
So the cost is not in performance, but in ease of understanding for the reader.

Reply:
I disagree with the claim that this is a cost to the reader. It's quite a trivial change. I do buy that signature verification algorithms are not commonly constant-time. However, we also don't have much experience with hybrid signature schemes, so assuming whatever we did before maps perfectly to what we should do here now doesn't seem right.

Food for thought.

Reply:
Yes, it's good to stay critical of these details!